php/libraries/Candidate.class.inc

```diff
@@ -699,52 +699,92 @@ class Candidate
     }
 
     /**
+    * Generates a new id in the format defined by the settings
+    *
+    * @param string   $name                  The name to display on exception
+    * @param array    $settings              The xml settings
+    * @param array    $static_dict           Example: array("siteAbbrev"=>"DCC")
+    * @param function $is_colliding_callback `bool ($needle)`
+    *
+    * @return string The id
+    */
+    private static function _generateId(
+        $name,
+        $settings,
+        $static_dict,
+        $is_colliding_callback
+    ) {
+        $generator = Candidate::idGeneratorFromXMLConfig(
+            $settings,
+            $static_dict
+        );
+
+        $is_sequential = (
+            isset($settings["generation"])
+            && $settings["generation"] == "sequential"
+        );
+
+        $result = null;
+        if ($is_sequential) {
+            if ($generator->nextNonColliding($is_colliding_callback)) {
+                $result = $generator->cur();
+            }
+        } else {
+            do {
+                $generator->nextRand();
+                $result = $generator->cur();
+            } while (!is_null($result) && $is_colliding_callback($result));
+        }
+
+        if (is_null($result)) {
+            $attempts = $generator->getAttempts();
+            $seqCount = $generator->sequenceCount();
+            $error    = "Could not generate {$name} after {$attempts} attempts. "
+                . "All {$seqCount} possible {$name} have been generated. "
+                . "Please set {$name} generator definition "
+                . "to allow more {$name} to be generated. ";
+            if (!$is_sequential) {
+                $error .= "(expected number of candidates)^3 is recommended "
+                    . "for random generation.";
+            }
+            throw new Exception($error);
```

Comment:
I'll make that change. LorisException seems reasonable enough, even though we don't seem to do anything with it.